web-common/src/features/charts/render/vega-signals.ts

```diff
@@ -0,0 +1,29 @@
+/**
+ * Util methods for handling vega signals
+ */
+
+export function resolveSignalField(value: unknown, field: string) {
+  if (typeof value === "object" && value !== null) {
+    return value[field]?.[0];
+  }
+  return undefined;
+}
+
+export function resolveSignalTimeField(value: unknown) {
+  /**
+   * Time fields end with `_ts`
+   * We iterate over the keys of the object and return the first key that ends with `_ts`
+   * */
+  if (typeof value === "object" && value !== null) {
+    for (const key in value) {
+      if (key.endsWith("_ts")) {
+        const ts = resolveSignalField(value, key);
+
+        if (ts !== undefined) {
+          return new Date(ts);
+        }
+      }
+    }
+  }
+  return undefined;
```

Comment:
**Optimization Issue**: The functions 'resolveSignalField' and 'resolveSignalTimeField' may lead to inefficient data handling and potential performance bottlenecks due to repeated object property access and unnecessary creation of Date objects within loops. <br> **Fix**: Optimize the data handling by caching object properties and reducing the creation of new Date objects unless absolutely necessary. Consider restructuring the logic to minimize deep property access and loop computations. <br> **Code Suggestion**: 
 ```
 +  const fieldValue = value[field];
 +  if (typeof value === "object" and value !== null && fieldValue) {
 +    return fieldValue[0];
 +  }
 +  return undefined;
 +
 +  const tsValue = value[key];
 +  if (typeof value === "object" and value !== null) {
 +    for (const key in value) {
 +      if (key.endsWith("_ts") && tsValue) {
 +        return new Date(tsValue);
 +      }
 +    }
 +  }
 +  return undefined;
 ```